### PR TITLE
Don't set MSBuildAllProjects for MSBuild 16.0+

### PR DESCRIPTION
**Pull request by drewnoakes**
Microsoft/msbuild#1299 discusses a change whereby MSBuild prepends the newest project file path to `MSBuildAllProjects` automatically. Therefore, so long as consumers of this property only use this property for up-to-date checks, it's possible to reduce both the size of this property and corresponding allocations.

This reduction applies equally to consumers, such as for the case discussed in dotnet/project-system#3744.

For backwards compatibility the property is still set for MSBuild versions prior to 16.0.

See also Microsoft/msbuild#3605.

//cc: @jeffkl @rainersigwald @davkean

### CODE REVIEW

File: src/Assets/TestProjects/AppWithSharedProject/SharedProject/SharedProject.projitems [hunk 1]
```diff
@@ -1,7 +1,9 @@
 ﻿<?xml version="1.0" encoding="utf-8"?>
 <Project xmlns="http://schemas.microsoft.com/developer/msbuild/2003">
   <PropertyGroup>
-    <MSBuildAllProjects>$(MSBuildAllProjects);$(MSBuildThisFileFullPath)</MSBuildAllProjects>
+    <MSBuildAllProjects Condition=" '$(MSBuildVersion)' == '' Or '$(MSBuildVersion)' &lt; '16.0' ">
```
Comment (livarcocc) [inline src/Assets/TestProjects/AppWithSharedProject/SharedProject/SharedProject.projitems]: I am not sure if this has been discussed. But these files should not be used with MSBuild less than 16.0.

The CLI will require Dev16, which means if someone tries to use this in Dev15, they will have to do hacks to enable it and we don't support it.

And the CLI itself will carry MSBuild 16.

cc @rainersigwald

**Comment by drewnoakes** [comment]
> here was my commit that I hadn't had a change to submit yet

If it's safe to remove them outright then that's a better approach, and given your commit we can close this PR.

> I'm surprised you didn't have to update the unit test like I had to.

It's marked as `Skip`. It does look like it should be removed, as you did.

**Comment by jeffkl** [comment]
My commit is from August and I haven't had time to send a PR so feel free to merge this PR instead.

**Comment by drewnoakes** [comment]
In that case I'll remove the skipped unit test and remove the `< 16.0` checks altogether.

**Review by nguerrera — APPROVED** [review]

**Review by rainersigwald — COMMENTED** [review]
🎉

I closed https://github.com/dotnet/sdk/issues/1810 since this will make it super obsolete.

**Comment by davkean** [comment]
> The CLI will require Dev16, which means if someone tries to use this in Dev15, they will have to do hacks to enable it and we don't support it.

How is that enforced? ie What will prevent this from "just working" if I run MSBuild.exe (v15) against a project that only has 3.0 installed.

**Review by rainersigwald — APPROVED** [review]

**Comment by livarcocc** [comment]
@davkean the 3.0 CLI has a file that establishes 16.0 as the minimum MSBuild version it requires. This file is then honored by the SDK Resolver that msbuild uses to find the .NET Core SDK props/targets. Users would have to edit this file inside the SDK installation to get around that. At that point, I would say all bets are OFF.

**Comment by davkean** [comment]
👍

**Pull request merged**